crates/dojo-core/src/world.cairo

```diff
@@ -291,16 +302,23 @@ mod World {
     fn delete_entity(context: Context, component: ShortString, query: Query) {
         // Assert can only be called through the executor
         // This is to prevent system from writing to storage directly
-        assert(get_caller_address() == executor_dispatcher::read().contract_address, 'must be called thru executor');
+        assert(
+            get_caller_address() == executor_dispatcher::read().contract_address,
+            'must be called thru executor'
+        );
 
         // Get execution role
         let role = execution_role();
 
-        // Validate the calling system has permission to write to the component
-        assert(
-            is_authorized(context.caller_system, component, AuthRole { id: role }),
-            'system not authorized'
-        );
+        // Validate authorization if role is not set
+        // Otherwise, proceed with the write
+        if role.into() == 0 {
+            // Validate the calling system has permission to write to the component
+            assert(
+                is_authorized(context.caller_system, component, AuthRole { id: role }),
+                'system not authorized'
+            );
+        };
```

Comment:
we should deduplicate between here and set just since it is critical path and should be maintained in one place

Reply:
Will refactor this to an internal function